settings: Convert names to pills in organization settings.

palashb01 opened the pull request:
- [x] Convert the plan anchor tags to pills displaying the user's avatar with the proper user name.
- [x] Users, Bots, Bot owners, Deactivated users all are migrated to pills.
- [x] If the name is too long or the viewport becomes smaller, the user's name is trimmed and abbreviated with ellipsis `...`
- [x]  When the viewport goes below 'lg_min' (992px), the email column is dropped from all three tables: users, deactivated     
          users, and bots.
- [x] Show `no owner` as a plain text inside bots table.
- [x] Rename name column to `User`.
- [x] Adjust the column widths to improve the visibility of tables.
- [x] Change the color of the pills for the dark theme.
- [x] Change the default width of the user column from 100px to occupy 30% of the  space.
- [x] Add a tippy tooltip for name pills.

-------------

Fixes: #24229
CZO: [Thread](https://chat.zulip.org/#narrow/stream/101-design/topic/Convert.20name.20to.20avatar.20pills.20in.20settings.20.2324308)
for Rows issue: [Thread](https://chat.zulip.org/#narrow/stream/49-development-help/topic/Settings.20Table.20Rows.20CSS.20Transparency)

------------------

### Screenshots and screen captures:

<details>
<summary>Overall Changes:</summary>

![overall](https://user-images.githubusercontent.com/66828942/223624376-3a691ab7-e68e-45db-83aa-c4fa4f2131ea.gif)


</details>

<details>
<summary>User Pill:</summary>

| Light | Dark |
|---------|---------|
| ![Screenshot from 2023-03-08 10-30-39](https://user-images.githubusercontent.com/66828942/223624465-017f4579-5c50-47c6-ba2a-08315a5bb086.png) | ![Screenshot from 2023-03-08 10-30-26](https://user-images.githubusercontent.com/66828942/223624484-7b892bb7-f99b-4afb-968a-cc04bae79837.png) |



</details>

<details>
<summary>Ellipses Change:</summary>

<br>

![ellipsis_change](https://user-images.githubusercontent.com/66828942/223624607-19e7bc38-3713-4f71-8ca9-42387d61df5f.gif)


</details>

<details>
<summary>Users UI:</summary>

**Light:**

![Screenshot from 2023-03-08 10-32-45](https://user-images.githubusercontent.com/66828942/223624694-001e9f3c-ac6e-4454-987c-2de12eb03207.png)

<br/>

**Dark:**

![Screenshot from 2023-03-08 10-32-34](https://user-images.githubusercontent.com/66828942/223624711-48a7e22f-f157-4edf-b7ee-554aa77a915a.png)


</details>

<details>
<summary>Bots list:</summary>

**Light:**

 ![Screenshot from 2023-03-08 10-32-58](https://user-images.githubusercontent.com/66828942/223624813-062cbfe3-105d-4e35-88ee-604b70fbffc1.png)

<br/>

**Dark:**

![Screenshot from 2023-03-08 10-33-06](https://user-images.githubusercontent.com/66828942/223624841-8b44d4c9-522e-4869-974a-61ec08878be8.png)



</details>

<details>
<summary>View port changes:</summary>
<details>
<summary>Bots list:</summary>

![bots_change](https://user-images.githubusercontent.com/66828942/223625523-f5c2bdb7-9340-4fa2-98ff-a7cb52ce8b8f.gif)

</details>
<details>
<summary>Users list:</summary>

![users_change](https://user-images.githubusercontent.com/66828942/223625583-be251dbe-83ca-4eb2-bea9-13a68ec5d082.gif)


</details>
</details>

<details>
<summary>Tippy tooltip:</summary>
<br>

<img width="214" alt="tooltip_name_pills" src="https://github.com/zulip/zulip/assets/66828942/d62dfdd1-b617-4555-aab1-9b74b2f2bd87">

</details>


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

palashb01 commented:
Hey @alya, could you please review the PR and suggest if any changes are required, Thanks.

sov-j commented:
I have tested this on current Firefox (109.0.1). It works according to the screenshots and almost - to the description in #24229  except Bots panel. If I understand #24229 corectly, bot and owner names must have avatars too, but now they don't.

palashb01 commented:
> I have tested this on current Firefox (109.0.1). It works according to the screenshots and almost - to the description in #24229 except Bots panel. If I understand #24229 corectly, bot and owner names must have avatars too, but now they don't.

Hey @sov-j ,The code has been written such that if a bot avatar is available, it will display it. However, having an avatar for the bot owner could take up a lot of space, and since the space in the bot's panel is limited, I preferred not to render the avatar for bot owners.

palashb01 commented:
Hey @sov-j , the multiple columns in the bot-panel have limited space, which is causing an issue as the inclusion of both bot and bot owner's avatar is taking up too much space, resulting in the names being obscured on narrow screens due to the lack of space.

**This is how it looks on narrow_screen:**
![Screenshot from 2023-02-09 00-51-56](https://user-images.githubusercontent.com/66828942/217631441-84b18d51-9e46-4aac-940e-8d48f75cf2e8.png)

sov-j commented:
Sure, it makes to me, it is just not exactly what #24229 states as requirements, so as a tester I had to mention it :)

palashb01 commented:
Hey @alya @amanagr , could you please review the PR and suggest if any changes are required, Thanks.

palashb01 commented:
After the recent conversation within the [thread](https://chat.zulip.org/#narrow/stream/101-design/topic/Convert.20name.20to.20avatar.20pills.20in.20settings.20.2324308), the PR has been updated. This update involves removing the email column inside bot_type when the viewport is smaller than sm_min (575px), which allows for the implementation of both the avatar and name inside the pill.

palashb01 commented:
Hey @alya @amanagr , could you please review the PR and suggest if any changes are required, Thanks.

alya commented:
The pills do not look right in light theme:

![Screen Shot 2023-03-01 at 12 24 48 PM](https://user-images.githubusercontent.com/2090066/222257551-ad8aa463-4ed2-4aba-8181-68049ac92309.png)

Please post updated screenshots when this has been fixed, and be sure to test visual changes in both themes next time. See https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#manual-testing for detailed manual testing guidelines.

alya commented:
Other notes:

1. Clicking anywhere in the pill (not just the name) should open the user card.
2. The hover colors look quite bad in dark theme. We probably don't need to make the pill green on hover, and we definitely don't need to make the name within it blue.

![Screen Shot 2023-03-01 at 12 27 42 PM](https://user-images.githubusercontent.com/2090066/222258246-8611530c-7b95-4b35-8a91-ea126054254c.png)

alya commented:
The column sizes at various screen widths feel OK to me.

palashb01 commented:
Hey @alya , I have implemented all the changes suggested by you and updated the PR description with the latest changes in both light and dark mode. Could you please review the PR again? Thanks.

alya commented:
Thanks! This looks reasonable to me now, but I think it would be worth posting some screenshots in #design in case other folks have feedback.

The one tweak I noted is that we shouldn't show a hand cursor for "No owner" on the bots tab, since clicking on it doesn't do anything. I'm not sure if "No owner" works best as a pill or just plain text.

palashb01 commented:
> Thanks! This looks reasonable to me now, but I think it would be worth posting some screenshots in #design in case other folks have feedback.
> 
> The one tweak I noted is that we shouldn't show a hand cursor for "No owner" on the bots tab, since clicking on it doesn't do anything. I'm not sure if "No owner" works best as a pill or just plain text.

Dropped a message in the #design thread for both the suggestions. [Thread](https://chat.zulip.org/#narrow/stream/101-design/topic/Convert.20name.20to.20avatar.20pills.20in.20settings.20.2324308)

palashb01 commented:
Hey @alya , I have made all of the suggested changes and updated the description, could you please review it again? Thanks.

sahil839 commented on web/styles/settings.css:
```diff
@@ -1937,3 +1937,34 @@ $option_title_width: 180px;
+.pill_container {
+    display: -webkit-box;
+    -webkit-line-clamp: 1;
+    -webkit-box-orient: block-axis;
```
I can see that some of these properties are non-standard and may have some issues [here[(https://developer.mozilla.org/en-US/docs/Web/CSS/-webkit-line-clamp). Can we instead use flex layout? If there is no other way to make this work as expected, as we already use these properties in a couple of places, but we can try once.

palashb01 replied:
Changed this to use flex layout.

sahil839 reviewed (commented):
Posted few comments. A couple of points -
- Is there a reason the pills in these sections are different? And do we plan to make all the user pills consistent?
- We increased the role column width to fit the `Administrator` text, especially as it is quite long in a couple of languages other than English. In English also, it looks odd with current minimum width value 
![Screenshot from 2023-03-13 15-14-40](https://user-images.githubusercontent.com/35494118/224665279-4228dfc0-0ca8-4211-bef6-bdc56c1771bc.png)
- We can probably add tooltips when hovering hover the pills. It can be helpful to show long names.

palashb01 commented on web/src/settings_users.js:
```diff
@@ -259,8 +269,15 @@ function human_info(person) {
     info.can_modify = page_params.is_admin;
     info.is_current_user = people.is_my_user_id(person.user_id);
     info.cannot_deactivate = info.is_current_user || (person.is_owner && !page_params.is_owner);
+
```
Removed it.

palashb01 commented on web/src/settings_users.js:
```diff
@@ -218,6 +218,12 @@ function bot_info(bot_user_id) {
+    if (bot_user.avatar_url) {
+        info.img_src = bot_user.avatar_url;
+    } else {
+        info.img_src = people.medium_avatar_url_for_person(bot_user);
```
Simplified the if-else block and replaced `medium_avatar_url_for_person` with `small_avatar_url_for_person`. I had this block because the `user_object/bot_object` already had an `avatar_url` present, and in cases where this might be empty, I used `medium_avatar_url_for_person`.

palashb01 commented:
> Posted few comments. A couple of points -
> 
>     * Is there a reason the pills in these sections are different? And do we plan to make all the user pills consistent?
> 
>     * We increased the role column width to fit the `Administrator` text, especially as it is quite long in a couple of languages other than English. In English also, it looks odd with current minimum width value
>       ![Screenshot from 2023-03-13 15-14-40](https://user-images.githubusercontent.com/35494118/224665279-4228dfc0-0ca8-4211-bef6-bdc56c1771bc.png)
> 
>     * We can probably add tooltips when hovering hover the pills. It can be helpful to show long names.

- We can open a separate issue to simplify all the user pills throughout the app. There are many instances where we use different user pills, so it would be a good idea to have a separate issue for this migration.

- Recently, we have set a minimum width of 100px for every column in the table. This fixes the administrator issue you mentioned.

- Yes, we can add a tooltip for user pills. Would you like me to make that change in this PR or should I open a new PR? Additionally, I believe this can be done for every user pill throughout the app.

sahil839 commented on web/styles/settings.css:
```diff
@@ -396,8 +396,12 @@ td .button {
-.user_role {
-    min-width: 95px;
+.user_id {
+    min-width: 60px !important;
```
We generally avoid using `!important`. Would be better to instead use a more specific selector.

sahil839 commented on web/styles/settings.css:
```diff
@@ -396,8 +396,12 @@ td .button {
-.user_role {
-    min-width: 95px;
+.user_id {
+    min-width: 60px !important;
```
We generally avoid using `!important`. Would be better to instead use a more specific selector.

palashb01 replied:
I have replaced it with a new ID. I'm not certain if this approach is correct or not. Please let me know if you have a better approach for this. Another approach that comes to mind is modifying the `app_components.css` file to include a check using `:not` to exclude applying the min-width to this specific column. However, I chose not to modify the `app_components.css` and instead added a new ID called `user_id_column`.

sahil839 commented on web/styles/settings.css:
```diff
@@ -1893,6 +1893,13 @@ $option_title_width: 180px;
+    .active,
+    .user_name_column {
+        position: sticky;
+        z-index: 2 !important;
```
Here also you can avoid using `important` by using a more specific selector.

palashb01 replied:
Same reason as above.

sahil839 reviewed (commented):
Posted some comments about the CSS changes.

I cannot think of any solution for the narrow screen issue that you discussed on CZO. Would update if I find one.

sahil839 commented:
> We can open a separate issue to simplify all the user pills throughout the app. There are many instances where we use different user pills, so it would be a good idea to have a separate issue for this migration.

I am not sure what version would be better, the one we have in this PR or the existing one. We can have more feedback on this.


> Yes, we can add a tooltip for user pills. Would you like me to make that change in this PR or should I open a new PR? Additionally, I believe this can be done for every user pill throughout the app.

Would be fine doing in this PR as well, and having a separate commit would be ideal. For other pills in the app, I am not sure. I think it will be helpful where we clip the names due to width restriction, but maybe not helpful for all cases.

palashb01 commented on web/styles/settings.css:
```diff
@@ -1945,3 +1945,42 @@ $option_title_width: 180px;
+
+    &:hover {
+        border: 1px solid hsl(0deg 0% 20%);
+        outline: none;
```
The border is for the light theme. However, I have removed the outline because it is unnecessary.

palashb01 commented on web/styles/settings.css:
```diff
@@ -1945,3 +1945,42 @@ $option_title_width: 180px;
+    }
+
+    .deactivated-user-icon {
+        display: flex;
```
sure, I have changed it to this:

<img width="102" alt="deactivated" src="https://github.com/zulip/zulip/assets/66828942/9b365cb8-626f-43d0-9da3-e8c8599ecd5f">

palashb01 commented on web/styles/settings.css:
```diff
@@ -1945,3 +1945,42 @@ $option_title_width: 180px;
+    word-break: break-all;
+    width: fit-content;
+    white-space: nowrap;
+    overflow: hidden;
```
Thank you for bringing that to my attention. I have removed those values. I'm not entirely sure why I included them. Perhaps it was an oversight while migrating from webkit to flex layout. In any case, they have been removed now.

palashb01 commented:
@sahil839  , I have implemented all the changes you suggested and added a tippy tooltip to display the names. I have also updated the pull request (PR) description with the latest changes and included screenshots.

sahil839 reviewed (commented):
Posted one comment. Otherwise changes LGTM.

Regarding the issue for sticky columns you mentioned on CZO, I cannot find any solution other than the one already suggested on CZO to actually change the colors of rows.

palashb01 commented on web/src/settings_users.js:
```diff
@@ -236,6 +237,10 @@ function bot_info(bot_user_id) {
     info.display_email = bot_user.email;
 
+    if (owner_id) {
+        info.owner_img_src = people.get_by_user_id(owner_id).avatar_url;
```
replaced it with `small_avatar_url_for_person`

palashb01 commented:
@sahil839 @karlstolley I have updated  the PR with the latest changes and the method you suggested, could you please review the PR, thanks :)
and this is the overall change after removing the alpha dependency, I have updated the CZO thread in the PR description for any review:

![rows](https://github.com/zulip/zulip/assets/66828942/a0e25a86-8b82-47e5-aa0e-90b71c8db361)

sahil839 commented:
The pill does not live-update correctly when I change a user's name (I have not tested by changing other settings) from the "Manage user" modal. Looks good to me otherwise.

And as I mentioned above, the pills added here are inconsistent with the ones that exist already. I also noticed, we do not show the user's status emoji in the pills added here. So, would be useful to make them consistent in future and also probably remove some duplicate CSS.

karlstolley commented:
I agree with @sahil839 here, quite strongly:

> And as I mentioned above, the pills added here are inconsistent with the ones that exist already. I also noticed, we do not show the user's status emoji in the pills added here. So, would be useful to make them consistent in future and also probably remove some duplicate CSS.

Would it not be possible to rework the `input_pill.hbs` file? I could even imagine making that into something like `user_pill.hbs` in a way that perhaps takes a variable for whether or not to display the X--which is needed in the compose box but not in this case.

palashb01 commented:
@karlstolley @sahil839 Thanks a lot for the feedback, I have started a [CZO Thread](https://chat.zulip.org/#narrow/stream/6-frontend/topic/User.20Pill.20in.20settings.20Panel) for further conversation on this, Thanks a lot :)

Joelute commented:
Hey @alya. This PR could be marked as close as the issue has been completed by another PR.

sahil839 commented:
Thanks for all the work on this @palashb01. Closing this since this completed in #30371.
